Examples/WindowsFormsApplication/Nfield.cs

```diff
@@ -0,0 +1,138 @@
+            client.UserTokenCache.SetAfterAccess(AfterAccessNotification);
+
+            return client;
+        }
+
+        private async Task<string> ProvideTokenAsync()
+        {
+            var result = await AuthenticateAsync((status) => { });
+            return result.AccessToken;
+        }
+
+        private static string CreateScope(string scope) => $"{ApplicationConfiguration.Current.NfieldApiApplicationId}/{scope}";
+
+        internal async Task LogoutAsync()
+        {
+            var accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            while (accounts.Any())
+            {
+                await ClientApp.RemoveAsync(accounts.First());
+                accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            }
+        }
+
+        private byte[] RestoreTokenCacheFromRegistry()
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                var value = key.GetValue("TokenCache");
+                if (value is byte[] state)
+                    return state;
+            }
+
+            return null;
+        }
+
+        private void SaveTokenCacheInRegistry(byte[] state)
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                key.SetValue("TokenCache", state, RegistryValueKind.Binary);
+            }
+        }
+
+        public void BeforeAccessNotification(TokenCacheNotificationArgs args)
+        {
+            args.TokenCache.DeserializeMsalV3(RestoreTokenCacheFromRegistry());
+        }
+
+        public void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            if (args.HasStateChanged)
+            {
+                SaveTokenCacheInRegistry(args.TokenCache.SerializeMsalV3());
+            }
+        }
+
+        public async Task<AuthenticationResult> TryAuthenticateSilentAsync()
+        {
+            var accounts = await ClientApp.GetAccountsAsync();
+            var account = accounts.FirstOrDefault();
+
+            try
+            {
+                return await ClientApp.AcquireTokenSilent(Scopes, account).ExecuteAsync();
+            }
+            catch (MsalUiRequiredException)
+            {
+                return null;
+            }
+        }
+
+        public async Task<AuthenticationResult> AuthenticateAsync(Action<string> statusCallback)
+        {
+            var accounts = await ClientApp.GetAccountsAsync();
+            var account = accounts.FirstOrDefault();
+
+            try
+            {
+                return await ClientApp.AcquireTokenSilent(Scopes, account).ExecuteAsync();
+            }
+            catch (MsalUiRequiredException mure)
+            {
+                statusCallback(mure.Message);
+            }
+
+            var authResult = await ClientApp.AcquireTokenInteractive(Scopes)
```

Comment:
This kind of stuff definately can use some documentation (why are we doing interactive login at this point). If you get MsalUiRequiredException it usually means that e.g. consent should be given but that might not be fully clear to the user here without some context.

Reply:
It is more likely that you simply need to authenticate, I'll add comments.